answer2_Swift.playground/Contents.swift

```diff
@@ -0,0 +1,253 @@
+import Foundation
+//: # Question 2
+/*
+ Design and implement a data structure for a cache, which has the following functions:
+ 
+ get(key)
+ Return the value associated with the specified key if it exists in the cache, else return -1.
+ 
+ put(key, value, weight)
+ Associate value with key in the cache, such that value might be later retrieved by get(key).
+ 
+ Cache has a fixed capacity, and when such capacity is reached, key with the least score must be invalidated until the number of key s falls within cache capacity. The score is calculated as follows:
+            
+        weight ／ [ln(current_time - last_accessed_time + 1) + 1]
+ 
+ Implementation of the cache should be optimized on the time complexity of get(key).
+ For example, the average time complexity of get(key) could be constant.
+ For the purpose of implementing the cache, you could assume that common data structures, such as arrays, different types of lists, and hash tables, are available.
+ At the end of the answer, give and explain the computational complexity of get(key) and put(...) in Big O notation.
+ */
+//:----
+//: ### Analyze
+/*
+ Question 2 is a variant of the LRU Cache.
+ 
+ Step0: Complete LRU Cache
+ */
+//:----
+//: ## LRU Cache
+//: ### Analyze
+/*
+ LRU Cache retain a sorted and finite amount of data. The data is sorted in the order by the time it been accessed.
+ When the capacity is full, the least recently used object of data will be delete.
+ 
+ Step0: A payload protocol to contain key and value for node.
+ Step1: CachePayload for node to store.
+ Step2: Node contain payload, previous link and next link.
+ Step3: Doubly Linked List insert, remove, MoveTo
+ */
+//:----
+protocol Payload {
+    associatedtype Key
+    associatedtype Value
+    
+    var key: Key { get set }
+    var value: Value { get set }
+}
+
+struct CachePayload<T: Hashable, U>: Payload {
+    
+    var key: T
+    var value: U
+    
+    init(key: Key, value: Value) {
+        self.key = key
+        self.value = value
+    }
+}
+
+class Node<T: Payload> {
+    
+    var payload: T
+    var previous: Node<T>?
+    var next: Node<T>?
+    
+    init(value: T) {
+        self.payload = value
+    }
+}
+
+class DoublyLinkedList<T: Payload> {
+    
+    // The head starting at index 0.
+    var head: Node<T>?
+    var tail: Node<T>?
+    var isEmpty: Bool {
+        return head == nil && tail == nil
+    }
+    
+    public var count: Int = 0
+    
+    public func insertHead(node: Node<T>) {
+        guard !isEmpty else {
+            head = node
+            tail = node
+            count += 1
+            return
+        }
+        
+        var currNode = head
+        
+        if currNode === head {
+            if head === tail {
+                node.next = head
+                head?.previous = node
+                head = node
+            } else {
+                node.next = head
+                head = node
+            }
+            
```

Comment:
node 現在是 head，node 前要清空
```suggestion
            node.previous = nil
```